### ISSUE TITLE
`invalid-bool-returned` error only for case when `__bool__()` returns a constant

### ISSUE THREAD

**Issue by matusvalo** [PyCQA/pylint#6627]
### Current problem

Currently pylint can raise `invalid-bool-returned` error only for case when `__bool__()` returns a constant:
```python
class Rectangle1:

    def __init__(self, width, length):
        self.width = width
        self.length = length

    def __bool__(self):
        return 'WrongValue`  # [invalid-bool-returned]
```

This is pretty narrow check since mostly `__bool__` contains some kind of logic - e.g.:

```python
class Rectangle2:

    def __init__(self, width, length):
        self.width = width
        self.length = length

    def __bool__(self):
        return True if self.width == 0 and self.length == 0 else False
```

E.g. Following example should rise the error but no error is raised:

```python
class Rectangle3:

    def __init__(self, width, length):
        self.width = width
        self.length = length

    def __bool__(self):
        if self.width == 1:
            return 'Hello'
        if self.width == 2:
            return 3
        if self.width == 3:
            return 4.0
        if self.width == 4:
            return b'World'
        if self.width == 5:
            return None
        if self.width == 6:
            return True
        if self.width == 7:
            return object()
        if self.width == 8:
            return object
```

The behaviour is caused by following code which limits that `__bool__` function should have only one result:

https://github.com/PyCQA/pylint/blob/a57fb82fb40b6f38f156bab1b06aa30a833c6046/pylint/checkers/classes/special_methods_checker.py#L29-L38

The `Rectangle3` example has more than 1 results:

```python
(Pdb) p list(node.infer_call_result(caller, context=context))
[<Const.str l.9 at 0x10af097c0>, <Const.int l.11 at 0x10af09970>, <Const.float l.13 at 0x10af095b0>, <Const.bytes l.15 at 0x10af09460>, <Const.NoneType l.17 at 0x10af09310>, <Const.bool l.19 at 0x10af091c0>, <Instance of builtins.object at 0x4479109344>, <ClassDef.object l.0 at 0x109d4f100>]
```

### Desired solution

Instead of limiting `__bool__()` one inferred node in https://github.com/PyCQA/pylint/blob/a57fb82fb40b6f38f156bab1b06aa30a833c6046/pylint/checkers/classes/special_methods_checker.py#L23

We should check whether all inferred values are bool - e.g.:
```python
if all(isinstance(n,  Const.bool) for n in node.infer_call_result(caller, context=context)):
    # Check succeded - all return values are bools
else:
    # Check failed - at least one value is not bool
```

Note: Of course the proposed "solution" is just illustrative high level sketch code which should be made more robust

Question: Does the proposal make sense? Is the proposed solution possible fix or I have missed something?

### Additional context

The similar fix can be done for all other checks validating return value of special methods - `invalid-str-returned`, `invalid-repr-returned`, etc...

**Comment by matusvalo** [comment]
I have additional question. Following examples:
```python
class Rectangle1:

    def __init__(self, width, length):
        self.width = int(width)
        self.length = int(length)

    def __bool__(self):
        return self.width == 0 and self.length == 0

class Rectangle2:

    def __init__(self, width: int, length: int):
        self.width: int = int(width)
        self.length: int = int(length)

    def __bool__(self):
        return self.width == 0 and self.length == 0
```

For both of them, `infer_call_result` returns `Uninferable` result.

```python
(Pdb) p list(node.infer_call_result(caller, context=context))
[Uninferable]
```

I am not expert in astroid and dummy in inference but from my nonexpert point of view both cases have pretty deterministic  return value type. Or am I missing something?
